File: android/engine/src/main/kotlin/dev/timeframe/engine/ingest/Ingest.kt
package dev.timeframe.engine.ingest

import dev.timeframe.engine.ports.SessionRow
import java.time.Instant

object Ingest {
    fun validate(session: SessionRow): Boolean {
        if (session.id.isEmpty()) return false
        if (session.source != "chrome_extension" && session.source != "android") return false
        if (session.device != "computer" && session.device != "phone") return false
        if (session.label.isEmpty()) return false
        if (session.seconds <= 0) return false
        return try {
            Instant.parse(session.started_at)
            session.ended_at?.let { Instant.parse(it) }
            true
        } catch (_: Exception) {
            false
        }
    }

    fun dedupeAndGuard(rows: List<SessionRow>): List<SessionRow> {
        val seen = mutableSetOf<String>()
        val result = mutableListOf<SessionRow>()

        for (row in rows) {
            if (!validate(row)) continue
            if (seen.contains(row.id)) continue
            seen.add(row.id)

            val startMs = Instant.parse(row.started_at).toEpochMilli()
            val endMs = row.ended_at?.let { Instant.parse(it).toEpochMilli() } ?: (startMs + row.seconds * 1000)

            var sec = row.seconds
            val delta = Math.abs(endMs - startMs - sec * 1000)
            if (delta > 1000) {
                sec = Math.round((endMs - startMs) / 1000.0)
            }

            result.add(
                row.copy(
                    seconds = sec,
                    minutes = Math.round(sec / 60.0 * 100.0) / 100.0,
                    timezone = row.timezone ?: "Asia/Kolkata",
                    ended_at = row.ended_at ?: Instant.ofEpochMilli(endMs).toString()
                )
            )
        }
        return result
    }
}

File: web/src/lib/ingest.ts
import { EnrichedSession, Envelope, Session } from "./types";
import { cleanSessions } from "./clean";

/**
 * Hand-rolled session validator (no Zod per spec).
 */
export function validateSession(obj: unknown): obj is Session {
  if (!obj || typeof obj !== "object") return false;
  const s = obj as Record<string, unknown>;

  if (typeof s.id !== "string" || s.id.length === 0) return false;
  if (s.source !== "chrome_extension" && s.source !== "android") return false;
  if (s.device !== "computer" && s.device !== "phone") return false;
  if (typeof s.label !== "string" || s.label.length === 0) return false;
  if (typeof s.started_at !== "string") return false;
  if (typeof s.seconds !== "number" || isNaN(s.seconds) || s.seconds <= 0) return false;

  const startMs = Date.parse(s.started_at);
  if (isNaN(startMs)) return false;

  if (s.ended_at !== null && s.ended_at !== undefined) {
    if (typeof s.ended_at !== "string") return false;
    const endMs = Date.parse(s.ended_at);
    if (isNaN(endMs)) return false;
  }

  return true;
}

/**
 * Normalizes ISO timestamp with support for Z and +00:00.
 */
export function parseTimestampMs(iso: string): number {
  return new Date(iso).getTime();
}

/**
 * Deduplicates sessions by id, first occurrence wins.
 * Applies duration guard (|ended_at - started_at - seconds| > 1s -> recompute seconds).
 * Ensures timezone defaults to Asia/Kolkata when missing.
 */
export function ingestSessions(rawSessions: unknown[]): Session[] {
  const seenIds = new Set<string>();
  const validSessions: Session[] = [];

  for (const item of rawSessions) {
    if (!validateSession(item)) {
      continue;
    }

    if (seenIds.has(item.id)) {
      continue; // First occurrence wins
    }
    seenIds.add(item.id);

    const startMs = parseTimestampMs(item.started_at);
    let endMs = item.ended_at ? parseTimestampMs(item.ended_at) : startMs + item.seconds * 1000;

    let seconds = item.seconds;
    const deltaMs = Math.abs(endMs - startMs - seconds * 1000);
    if (deltaMs > 1000) {
      seconds = Math.round((endMs - startMs) / 1000);
    }

    const minutes = parseFloat((seconds / 60).toFixed(2));
    const timezone = item.timezone || "Asia/Kolkata";

    validSessions.push({
      ...item,
      seconds,
      minutes,
      timezone,
      ended_at: item.ended_at || new Date(endMs).toISOString(),
    });
  }

  return validSessions;
}

/**
 * Pure function: Envelope -> EnrichedSession[]
 */
export function buildLedger(
  envelope: Envelope,
  userPins?: { work?: string[]; killers?: string[] },
  userOverrides?: Record<string, import("./types").Category>
): EnrichedSession[] {
  const ingested = ingestSessions(envelope.sessions);
  return cleanSessions(ingested, userPins, userOverrides);
}
